unstructured/partition/ocr.py

```diff
@@ -196,19 +212,158 @@ def supplement_page_layout_with_ocr(
                         padded_element.bbox.y2,
                     ),
                 )
+                # Note(yuming): instead of getting OCR layout, we just need
+                # the text extraced from OCR for individual elements
                 text_from_ocr = get_ocr_text_from_image(
                     cropped_image,
                     ocr_languages=ocr_languages,
-                    entire_page_ocr=entire_page_ocr,
+                    ocr_agent=ocr_agent,
                 )
                 element.text = text_from_ocr
-        return page_layout
     else:
         raise ValueError(
             "Invalid OCR mode. Parameter `ocr_mode` "
             "must be set to `entire_page` or `individual_blocks`.",
         )
 
+    # Note(yuming): use the OCR data from entire page OCR for table extraction
+    if infer_table_structure:
+        if ocr_layout is None:
+            # Note(yuming): ocr_layout is None for individual_blocks ocr_mode
+            ocr_layout = get_ocr_layout_from_image(
+                image,
+                ocr_languages=ocr_languages,
+                ocr_agent=ocr_agent,
+            )
+        page_layout.elements[:] = supplement_element_with_table_extraction(
+            elements=page_layout.elements,
+            ocr_layout=ocr_layout,
+            image=image,
+        )
+
+    return page_layout
+
+
+def supplement_element_with_table_extraction(
+    elements: List[LayoutElement],
+    ocr_layout: List[TextRegion],
+    image: PILImage,
+) -> List[LayoutElement]:
+    """Supplement the existing layout with table extraction. Any Table elements
+    that are extracted will have a metadata field "text_as_html" where
+    the table's text content is rendered into an html string.
+    """
+    for element in elements:
+        if element.type == "Table":
+            padding = env_config.IMAGE_CROP_PAD
+            padded_element = pad_element_bboxes(element, padding=padding)
+            cropped_image = image.crop(
+                (
+                    padded_element.bbox.x1,
+                    padded_element.bbox.y1,
+                    padded_element.bbox.x2,
+                    padded_element.bbox.y2,
+                ),
+            )
+            table_tokens = get_table_tokens_per_element(
+                padded_element,
+                ocr_layout,
+            )
+            table_agent = init_table_agent()
+            element.text_as_html = table_agent.predict(cropped_image, ocr_tokens=table_tokens)
+    return elements
```

Comment:
What do you think about moving `table_agent = init_table_agent()` to the first line (out of the for `loop`) and then not defining `table_agent` as a global variable?

Reply:
or adding `if table_agent is None:` to `init_table_agent()`?
```
def init_table_agent():
    """Initialize a table agent from unstructured_inference as
    a global variable to ensure that we only load it once."""
    from unstructured_inference.models import tables

    global table_agent

    if table_agent is None:
        table_agent = tables.UnstructuredTableTransformerModel()
        table_agent.initialize(model="microsoft/table-transformer-structure-recognition")

    return table_agent
```

Reply:
good catch! i moved `init_table_agent` outside the loop and added `is None` check to it